### ISSUE TITLE
Project roadmap, documentation, samples, config etc

### ISSUE THREAD

**Issue by simon-st** [awslabs/smithy-kotlin#738]
I spent some time playing around with the smithy-kotlin-codegen. I realize it is still in an early stage, but I wanted to list some comments, questions and requests from the perspective of someone that tried to use the project without prior knowledge:

- What is the roadmap for this project? Do you have any idea when you will get close to a v1 release?
- No documentation. I had to figure out by trial and error how to use this plugin. It took me forever to even find the plugin name to use in the smithy-build.json. Just later, I found [this](https://github.com/awslabs/aws-sdk-kotlin/blob/main/docs/howto/client-generator.md) documentation in a different repo that gives at least some information. It would be great if basic setup and use of the plugin was documented right here in this repo.
- Additional documentation would be appreciated as well. E.g. in what ways can the plugin be configured?
- A sample project would be nice to get started
- This might better be it's own ticket, but I was wondering if there is a way to configure the output classes. Specifically I would prefer the models to be data classes over the current builder pattern (or maybe both)

**Comment by simon-st** [comment]
Thanks for the extensive explanation. The linked discussion about data classes vs builder is very insightful. 

We are currently still evaluating what codegen to use. We are mostly interested in generating the models for the client, so Smithy might be too much overhead since it requires a service plus protocol. 

The most important feedback I wanted to give is that, without any prior knowledge of Smithy or the AWS services, it is very hard to get started. It's a big, complex topic and the documentation seems insufficient. Again, I understand that you are still in the midst of development. I will keep an eye on the project for sure to see where it goes.

**Comment by ianbotsf** [comment]
> We are mostly interested in generating the models for the client, so Smithy might be too much overhead since it requires a service plus protocol.

Do you mean you're primarily interested in generating data structures from the model but not necessarily any serialization/deserialization for transmitting them to/from services?

**Comment by simon-st** [comment]
Correct. For now at least. I will need codegen for kotlin, swift and typescript. The Smithy plugins for those languages don't seem to be ready for primetime outside of AWS services yet from how I see that. If I can just generate the models for now though, that would already help. Our backend team is providing smithy files that they use internally and I was hoping that I can use those for the client models as well.

**Comment by aajtodd** [comment]
> since it requires a service plus protocol.

It requires a service but without a protocol it will spit out basically just a service definition + models. The reason it requires a service is that the code generator works off the concept of a "Shape closure", i.e. all of the shapes reachable within some defined closure. For client generation this is the service shape. 

There isn't really a path defined to generate just models at this point (e.g. like protobuf). It would certainly be a good feature though. If the backend smithy models have a service shape then you could probably do what you want.
